src/mesh/mesh_dual_weighted_residual_error.cpp

```diff
@@ -0,0 +1,388 @@
+    dg->high_order_grid->execute_coarsening_and_refinement();
+
+    dg->allocate_system();
+    dg->solution.zero_out_ghosts();
+
+    dg->solution = solution_coarse;
+
+    adjoint_state = AdjointStateEnum::coarse;
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+dealii::LinearAlgebra::distributed::Vector<real> DualWeightedResidualError<dim, nstate, real, MeshType>::fine_grid_adjoint(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    convert_to_state(AdjointStateEnum::fine, dg);
+
+    // dIdw_fine.reinit(dg->solution);
+    // dIdw_fine = functional.evaluate_dIdw(dg, physics);
+    const bool compute_dIdW = true, compute_dIdX = false;
+    const real functional_value = functional->evaluate_functional(compute_dIdW,compute_dIdX);
+    (void) functional_value;
+    dIdw_fine = functional->dIdw;
+
+    adjoint_fine.reinit(dg->solution);
+    
+    dg->assemble_residual(true);
+    dg->system_matrix *= -1.0;
+
+    dealii::TrilinosWrappers::SparseMatrix system_matrix_transpose;
+    Epetra_CrsMatrix *system_matrix_transpose_tril;
+
+    Epetra_RowMatrixTransposer epmt(const_cast<Epetra_CrsMatrix *>(&dg->system_matrix.trilinos_matrix()));
+    epmt.CreateTranspose(false, system_matrix_transpose_tril);
+    system_matrix_transpose.reinit(*system_matrix_transpose_tril,true);
+    delete system_matrix_transpose_tril;
+    solve_linear(system_matrix_transpose, dIdw_fine, adjoint_fine, dg->all_parameters->linear_solver_param);
+    // solve_linear(dg.system_matrix, dIdw_fine, adjoint_fine, dg.all_parameters->linear_solver_param);
+
+    return adjoint_fine;
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+dealii::LinearAlgebra::distributed::Vector<real> DualWeightedResidualError<dim, nstate, real, MeshType>::coarse_grid_adjoint(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    convert_to_state(AdjointStateEnum::coarse, dg);
+
+    dIdw_coarse.reinit(dg->solution);
+    //dIdw_coarse = functional.evaluate_dIdw(dg, physics);
+    const bool compute_dIdW = true, compute_dIdX = false;
+    const real functional_value = functional->evaluate_functional(compute_dIdW,compute_dIdX);
+    (void) functional_value;
+    dIdw_coarse = functional->dIdw;
+
+    adjoint_coarse.reinit(dg->solution);
+
+    dg->assemble_residual(true);
+    dg->system_matrix *= -1.0;
+
+    dealii::TrilinosWrappers::SparseMatrix system_matrix_transpose;
+    Epetra_CrsMatrix *system_matrix_transpose_tril;
+
+    Epetra_RowMatrixTransposer epmt(const_cast<Epetra_CrsMatrix *>(&dg->system_matrix.trilinos_matrix()));
+    epmt.CreateTranspose(false, system_matrix_transpose_tril);
+    system_matrix_transpose.reinit(*system_matrix_transpose_tril);
+    solve_linear(system_matrix_transpose, dIdw_coarse, adjoint_coarse, dg->all_parameters->linear_solver_param);
+    // solve_linear(dg->system_matrix, dIdw_coarse, adjoint_coarse, dg->all_parameters->linear_solver_param);
+
+    return adjoint_coarse;
```

Comment:
All this code is exactly the same as the function above. Move common code into a function, and pass the different arguments to it.

Reply:
Moved common code to `compute_adjoint()`.